### PR TITLE
normalized fee perf improvement

### PR DESCRIPTION
**Pull request by isaacJChen**

### CODE REVIEW

File: lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts [hunk 1]
```diff
@@ -21,19 +26,46 @@ export default class NormalizedFeeCalculator implements IFeeCalculator {
...
-    // https://github.com/decentralized-identity/sidetree/issues/936
+    const shouldCalculate = this.shouldCalculateNormalizedFee(block);
+    if (!shouldCalculate.shouldCalculate) {
+      return shouldCalculate.fee!;
+    }
+    const blocksToAverage = await this.getLookBackBlocks(block);
+    return this.calculateNormalizedFee(blocksToAverage);
+  }
+
+  private shouldCalculateNormalizedFee (block: number): { shouldCalculate: boolean, fee: number | undefined } {
```
Comment (isaacJChen) [inline lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts:60]: Please give me a better name (cry)

File: lib/bitcoin/interfaces/IFeeCalculator.ts [hunk 2]
```diff
@@ -6,4 +9,9 @@ export default interface IFeeCalculator {
    * Returns the fee for a particular block height.
    */
   getNormalizedFee (block: number): Promise<number>;
+
+  /**
+   * Returns the block with normalized fee added
+   */
+  addNormalizedFeeToBlock (blockMetadataWithoutFee: BlockMetadataWithoutNormalizedFee): Promise<BlockMetadata>;
```
Comment (thehenrytsai) [inline lib/bitcoin/interfaces/IFeeCalculator.ts:16]: Naming nit:
```suggestion
  addNormalizedFeeToBlockMetadata (blockMetadataWithoutFee: BlockMetadataWithoutNormalizedFee): Promise<BlockMetadata>;
```

File: lib/bitcoin/versions/0.10.0/NormalizedFeeCalculator.ts [hunk 3]
```diff
@@ -1,10 +1,14 @@
+import BlockMetadata from '../../models/BlockMetadata';
```
Comment (thehenrytsai) [inline lib/bitcoin/versions/0.10.0/NormalizedFeeCalculator.ts:1]: Just a thought, should we detroy 0.10.0 folder altogether?

File: lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts [hunk 4]
```diff
@@ -21,19 +26,46 @@ export default class NormalizedFeeCalculator implements IFeeCalculator {
...
-      return this.initialNormalizedFee;
+      return { shouldCalculate: false, fee: this.initialNormalizedFee };
     }
-
-    const blocksToAverage = await this.getLookBackBlocks(block);
-    return this.calculateNormalizedFee(blocksToAverage);
+    return { shouldCalculate: true, fee: undefined };
   }
 
   private async getLookBackBlocks (block: number): Promise<BlockMetadata[]> {
```
Comment (thehenrytsai) [inline lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts:71]: Naming nit:
```suggestion
  private async getBlocksInLookBackWindow (block: number): Promise<BlockMetadata[]> {
```

File: lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts (hunk 1 above)
Comment (thehenrytsai) [reply lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts:60]: LOL. I'd suggest eliminating this method, and just implement the logic in both callers. I double checked, you don't need to handle pre-genesis case in both cases (`getNormalizedFee()` in `BitcoinProcessor` already does the same check and actually throws error which I think is the better behavior than returning 0), so code there would be more or less the same even without this method.

**Review by thehenrytsai — CHANGES_REQUESTED** [review]
Minor changes requested.

File: lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts (hunk 1 above)
Comment (isaacJChen) [reply lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts:60]: Oh, true. Ok.

**Review by thehenrytsai — APPROVED** [review]

**Pull request merged**